### PR TITLE
Google-cloud-pub-sub: dedicated connection pool; `withContext`

### PR DESCRIPTION
**Pull request by btomala**
## Purpose

This PR introduces cached host connection pool for google cloud pub-sub HTTP integration.

## References

References  #1958

## Changes

* google-cloud-pub-sub start using cashed host connection pool instead of shared one
* replace `GooglePubSubSource` with `Source.tick`
* subscribe source from pub-sub is Cancellable (you can cancel it during graceful shutdown process)
* binary compatibility is broken
* all public methods takes `HttpsConnectionContext` as a parameter, I need to do this for testing purposes, not sure if this is a correct approach.

## Background Context

Previously the google-cloud-pub-sub user single request to fetch/push data to pub-sub. Single request use shared connections pool, for performance reasons makes sense to use separate connection pool for pub-sub only. 
Using streaming akka-http API we have Cancellation implemented in the tick source. It is easier to understand than `GooglePubSubSource`. However, if someone writes it maybe there is something special what I missed.
We can also add context passing to correlate publishing requests with responses.
I decided to pass `HttpsConnectionContex` as one of the params because I don't know how to skip validation SSL config in the test. Looks like `akka.ssl-config.loose.acceptAnyCertificate = true` flag doesn't work in this case.

If you like this change let me know, then I'll update documentation.

### CODE REVIEW

File: google-cloud-pub-sub/src/test/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApiSpec.scala [hunk 1]
```diff
@@ -30,17 +47,28 @@ class PubSubApiSpec extends FlatSpec with BeforeAndAfterAll with ScalaFutures wi
   implicit val defaultPatience =
     PatienceConfig(timeout = 5.seconds, interval = 100.millis)
 
+  val sslContext: SSLContext = SSLContext.getInstance("TLS")
+  sslContext.init(null, Array(new NoopTrustManager()), null)
+
+  val hcc = Some(ConnectionContext.https(sslContext))
```
Comment (ennru) [inline google-cloud-pub-sub/src/test/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApiSpec.scala]: You replaced the trust manager, that switches off much more than the hostname verification.
Instead of passing the context around, set it on the HTTP extension via `Http().setDefaultClientHttpsContext(ConnectionContext.https(sslContext))`.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala [hunk 2]
```diff
@@ -121,87 +128,138 @@ private[pubsub] trait PubSubApi {
...
+  def pull[T](project: String,
+              subscription: String,
+              returnImmediately: Boolean,
+              maxMessages: Int,
+              parallelism: Int,
+              hcc: Option[HttpsConnectionContext] = None)(
       implicit as: ActorSystem,
       materializer: Materializer
-  ): Future[PullResponse] = {
+  ): Flow[(Done, Option[String], T), (Future[PullResponse], T), NotUsed] = {
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala]: The code doesn't use the context in `T` right now. The `pull` implementation is only used with `subscribe`, so I think we do not need it here.
If there is a use case for passing context, the `mapAsyncUnordered` should become a `mapAsync` as it could reorder elements which `FlowWithContext` shouldn't allow.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala [hunk 3]
```diff
@@ -27,61 +28,46 @@ protected[pubsub] trait GooglePubSub {
...
-    if (httpApi.isEmulated) {
-      Flow[PublishRequest].mapAsyncUnordered(parallelism) { request =>
-        httpApi.publish(config.projectId, topic, maybeAccessToken = None, request)
-      }
-    } else {
-      Flow[PublishRequest].mapAsyncUnordered(parallelism) { request =>
-        config.session.getToken().flatMap { accessToken =>
-          httpApi.publish(config.projectId, topic, Some(accessToken), request)
-        }
+  ): Flow[PublishRequest, immutable.Seq[String], NotUsed] =
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala]: `publish` would benefit from a `FlowWithContext` variant.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala [hunk 4]
```diff
@@ -27,61 +28,46 @@ protected[pubsub] trait GooglePubSub {
...
+    httpApi
+      .accessToken[PublishRequest](config, parallelism)
+      .via(httpApi.publish(config.projectId, topic, parallelism))
+      .mapAsyncUnordered(parallelism) {
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala]: This should be a `mapAsync(1)` to be sure the response order is kept.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala [hunk 5]
```diff
@@ -27,61 +28,46 @@ protected[pubsub] trait GooglePubSub {
...
       )
-    )
-
-  def acknowledge(
-      subscription: String,
-      config: PubSubConfig,
-      parallelism: Int = 1
-  )(implicit actorSystem: ActorSystem, materializer: Materializer): Sink[AcknowledgeRequest, Future[Done]] = {
-    import materializer.executionContext
+      .mapAsyncUnordered(parallelism) {
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala]: This should be a `mapAsync(1)` to be sure the response order is kept.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala [hunk 6]
```diff
@@ -121,87 +128,138 @@ private[pubsub] trait PubSubApi {
       AcknowledgeRequest(json.asJsObject.fields("ackIds").convertTo[immutable.Seq[String]]: _*)
     def write(ar: AcknowledgeRequest): JsValue = JsObject("ackIds" -> ar.ackIds.toJson)
   }
-  private implicit val pullRequestFormat = DefaultJsonProtocol.jsonFormat2(PubSubApi.PullRequest)
+  private implicit val pullRequestFormat = DefaultJsonProtocol.jsonFormat2(PullRequest)
+
+  private def pool[T](hcc: Option[HttpsConnectionContext])(
+      implicit as: ActorSystem
+  ): Flow[(HttpRequest, T), (Try[HttpResponse], T), Http.HostConnectionPool] =
+    if (isEmulated) {
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala]: Very nice to have the emulation pushed to a central place instead.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala [hunk 7]
```diff
@@ -27,61 +28,46 @@ protected[pubsub] trait GooglePubSub {
...
-        subscription = subscription,
-        returnImmediately = config.pullReturnImmediately,
-        maxMessages = config.pullMaxMessagesPerInternalBatch,
-        httpApi = httpApi
+  /**
+   * Creates a source pulling messages from subscription
+   */
+  def subscribe(subscription: String, config: PubSubConfig, parallelism: Int = 1)(
+      implicit actorSystem: ActorSystem,
+      materializer: Materializer
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala]: While changing APIs... 
This could make use of [Source.setup](https://doc.akka.io/docs/akka/current/stream/operators/Source-or-Flow/setup.html) to not require the Actor System and Materializer.
But it might be better to keep that in a different PR.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala [hunk 8]
```diff
@@ -27,61 +28,46 @@ protected[pubsub] trait GooglePubSub {
...
+  ): Source[ReceivedMessage, Cancellable] =
+    Source
+      .tick(0.seconds, 1.second, Done)
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala]: This is nice. It might happen that the requests behave a bit differently depending on the downstream demand than with the graph stage, but I don't see any major difference.

**Comment by btomala** [comment]
Hi, sorry for delay, I was on holiday. I'll review your comments soon.

File: google-cloud-pub-sub/src/test/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApiSpec.scala (hunk 1 above)
Comment (btomala) [reply google-cloud-pub-sub/src/test/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApiSpec.scala]: Switching off the host verification wasn't enough. That was the reason why I replace the trust manager.
I think I try it already and doesn't work. However, I'll give a try again.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala (hunk 2 above)
Comment (btomala) [reply google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala]: Yes, it is not used yet. I have a plan to use it. However, I want to limit number of changes in the first step. I'll expose it to the public API in next commit. I'll change `mapAsync` to unordered.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala (hunk 4 above)
Comment (btomala) [reply google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala]: I'll add context than this should stay as it is.

File: google-cloud-pub-sub/src/test/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApiSpec.scala (hunk 1 above)
Comment (btomala) [reply google-cloud-pub-sub/src/test/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApiSpec.scala]: That's works! I'll remove HttpsConnectionContext from API

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala (hunk 3 above)
Comment (btomala) [reply google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala]: Yes, I want to add context here.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala (hunk 4 above)
Comment (btomala) [reply google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala]: I make leave it only for `publish` and change to `mapAsync(1)` for others

**Comment by btomala** [comment]
Do you have milestone 2.0 roadmap? When do you have plan to release those changes? Do you have a milestone release?

**Comment by ennru** [comment]
We'll start releasing 2.0 milestones soon, snapshots are published after successfull builds of master https://doc.akka.io/docs/alpakka/current/other-docs/snapshots.html (which is sadly not often)

**Comment by ennru** [comment]
We'll cut a first 2.0.0 milestone this week https://github.com/akka/alpakka/issues/1998

**Comment by btomala** [comment]
Ok, thanks! Please check the public API after my last changes. If this is ok, I'll upgrade the documentation.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala (hunk 7 above)
Comment (ennru) [reply google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/scaladsl/GooglePubSub.scala]: Created https://github.com/akka/alpakka/issues/2017

**Comment by ennru** [comment]
The errors MiMa detected require you to add exclusion files. I just suggested a more comprehensive instruction for that in https://github.com/akka/alpakka/pull/2015/files

**Comment by btomala** [comment]
I rebased branch to master and add a file with excludes. Unfortunately, looks like it is not taken into consideration when filters are run. Don't know why.

**Comment by ennru** [comment]
The exclusions file should be `.../mima-filters/2.0.0-M1.backwards.excludes/pr-1983-use-dedicated-connection-pool.excludes`

**Comment by btomala** [comment]
Ok, thanks I didn't know that milestones are considered as binary compatible. I test it locally and it passed now.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala [hunk 9]
```diff
@@ -121,87 +129,138 @@ private[pubsub] trait PubSubApi {
...
+    val url: Uri = s"/v1/projects/$project/subscriptions/$subscription:pull"
+
+    Flow[(Done, Option[String])]
+      .mapAsync(1) {
+        case (_, maybeAccessToken) =>
+          Marshal((HttpMethods.POST, url, PullRequest(returnImmediately, maxMessages)))
+            .to[HttpRequest]
+            .map(request => (authorize(maybeAccessToken)(request), NotUsed))
+      }
+      .via(pool[NotUsed]())
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala]: Please make add a `pool` variant without context so that the using methods don't get to create strange tuples, such as 
```
  private def poolSimple()(
    implicit as: ActorSystem
  ): Flow[HttpRequest, Try[HttpResponse], Http.HostConnectionPool] =
    Flow[HttpRequest]
      .map(_ -> NotUsed)
      .viaMat(pool())(Keep.right)
      .map(_._1)
```

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala [hunk 10]
```diff
@@ -16,39 +16,47 @@ import akka.http.scaladsl.model.headers.OAuth2BearerToken
 import akka.http.scaladsl.unmarshalling.Unmarshal
 import akka.stream.Materializer
 import akka.stream.alpakka.googlecloud.pubsub._
-import spray.json._
+import akka.stream.scaladsl.FlowWithContext
+import akka.stream.scaladsl.Flow
+import akka.{Done, NotUsed}
 import spray.json.DefaultJsonProtocol._
-import spray.json.{deserializationError, DefaultJsonProtocol, JsObject, JsString, JsValue, RootJsonFormat}
+import spray.json.{DefaultJsonProtocol, JsObject, JsString, JsValue, RootJsonFormat, deserializationError, _}
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala]: `spray.json._` will do

File: google-cloud-pub-sub/src/test/java/docs/javadsl/ExampleUsageJava.java [hunk 11]
```diff
@@ -67,21 +67,38 @@ private static void example() throws NoSuchAlgorithmException, InvalidKeySpecExc
...
+
     // #publish-fast
     Source<PublishMessage, NotUsed> messageSource = Source.single(publishMessage);
     messageSource
         .groupedWithin(1000, Duration.ofMinutes(1))
-        .map(messages -> PublishRequest.create(messages))
-        .via(publishFlow)
+        .map(messages -> Pair.apply(PublishRequest.create(messages), context))
+        .via(publishFlowWithContext)
         .runWith(Sink.ignore(), materializer);
```
Comment (ennru) [inline google-cloud-pub-sub/src/test/java/docs/javadsl/ExampleUsageJava.java]: This example would be simpler without the context.

File: google-cloud-pub-sub/src/test/scala/docs/scaladsl/ExampleUsage.scala [hunk 12]
```diff
@@ -59,13 +59,34 @@ class ExampleUsage {
...
+  //#publish-single-with-context
+
   //#publish-fast
   val messageSource: Source[PublishMessage, NotUsed] = Source(List(publishMessage, publishMessage))
-  messageSource.groupedWithin(1000, 1.minute).map(PublishRequest.apply).via(publishFlow).to(Sink.seq)
+  messageSource
+    .groupedWithin(1000, 1.minute)
+    .map(grouped => PublishRequest(grouped) -> resultPromise)
+    .via(publishFlowWithContext)
+    .to(Sink.seq)
```
Comment (ennru) [inline google-cloud-pub-sub/src/test/scala/docs/scaladsl/ExampleUsage.scala]: Same here, the context doesn't seem to provide any value.

**Comment by btomala** [comment]
I made them simpler.

**Comment by ennru** [comment]
Thank you! Sorry if it grew bigger than you originally intended...
Great to have `withContext` support, it proves useful in many situations.

**Comment by btomala** [comment]
Thank you for your help in review!
